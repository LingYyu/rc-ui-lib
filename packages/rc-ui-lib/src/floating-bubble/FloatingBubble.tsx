import React, { useContext, useEffect, useMemo, useRef, useState } from 'react';
import classnames from 'classnames';
import { FloatingBubbleProps } from './PropsType';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import { renderToContainer } from '../utils/dom/renderToContainer';
import { Icon } from '../icon';
import getRect from '../hooks/get-rect';
import useWindowSize from '../hooks/use-window-size';
import { addUnit, pick } from '../utils';
import { useTouch } from '../hooks/use-touch';
import useEventListener from '../hooks/use-event-listener';
import { closest } from '../utils/closest';
import { useSetState } from '../hooks';

const FloatingBubble: React.FC<FloatingBubbleProps> = (props) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('floating-bubble', prefixCls);
  const { offset, axis, gap, magnetic, icon, onClick, teleport, children } = props;
  const windowSize = useWindowSize();
  const touch = useTouch();
  const rootRef = useRef<HTMLDivElement>();
  const prevX = useRef(0);
  const prevY = useRef(0);
  const [state, setState] = useSetState({
    x: 0,
    y: 0,
    width: 0,
    height: 0,
  });

  const [dragging, setDragging] = useState(false);
  const [show, setShow] = useState(true);
  const [initialized, setInitialized] = useState(false);
  const boundary = useMemo(() => {
    return {
      top: gap,
      right: windowSize.width - state.width - gap,
      bottom: windowSize.height - state.height - gap,
      left: gap,
    };
  }, [windowSize, gap, state]);

  const rootStyle = useMemo(() => {
    const style: React.CSSProperties = {};
    const x = addUnit(state.x);
    const y = addUnit(state.y);
    style.transform = `translate3d(${x}, ${y}, 0)`;
    if (dragging || !initialized) {
      style.transition = 'none';
    }
    return style;
  }, [state, dragging, initialized]);
  const updateState = () => {
    const { width, height } = getRect(rootRef.current);
    setState({
      x: offset.x > 1 ? offset.x : windowSize.width - width - gap,
      y: offset.y > 1 ? offset.y : windowSize.height - height - gap,
      width,
      height,
    });
  };
  const onTouchStart = (event) => {
    touch.start(event);
    setDragging(true);
    prevX.current = state.x;
    prevY.current = state.y;
  };
  const onTouchMove = (event) => {
    event.preventDefault();
    touch.move(event);
    if (axis === 'lock') return;
    if (!touch.isTap.current) {
      if (axis === 'x' || axis === 'xy') {
        let nextX = prevX.current + touch.deltaX.current;
        if (nextX < boundary.left) nextX = boundary.left;
        if (nextX > boundary.right) nextX = boundary.right;
        setState({ x: nextX });
      }
      if (axis === 'y' || axis === 'xy') {
        let nextY = prevY.current + touch.deltaY.current;
        if (nextY < boundary.top) nextY = boundary.top;
        if (nextY > boundary.bottom) nextY = boundary.bottom;
        setState({ y: nextY });
      }
    }
  };
  useEventListener('touchmove', onTouchMove, {
    target: rootRef.current,
  });
  const onTouchEnd = () => {
    setDragging(false);
    // setTimeout(() => {
    if (magnetic === 'x') {
      const nextX = closest([boundary.left, boundary.right], state.x);
      setState({ x: nextX });
    }
    if (magnetic === 'y') {
      const nextY = closest([boundary.top, boundary.bottom], state.y);
      setState({ y: nextY });
    }
    if (!touch.isTap.current) {
      const updateOffset = pick(state, ['x', 'y']);

      if (prevX.current !== updateOffset.x || prevY.current !== updateOffset.y) {
        props?.onOffsetChange?.(updateOffset);
      }
    }
    // }, 0);
  };
  useEffect(() => {
    // setShow(true);
    // setTimeout(() => {
    updateState();
    setInitialized(true);
    // }, 0);
    return () => {
      if (teleport) setShow(false);
    };
  }, []);

  const renderContent = () => {
    return (
      show && (
        <div
          className={classnames(bem())}
          ref={rootRef}
          onTouchStart={onTouchStart}
          onTouchEnd={onTouchEnd}
          onTouchCancel={onTouchEnd}
          onClick={(e) => {
            if (touch.isTap.current) {
              onClick?.(e);
            }
          }}
          style={rootStyle}
        >
          {children || <Icon name={icon} className={classnames(bem('icon'))} />}
        </div>
      )
    );
  };
  return teleport ? renderToContainer(teleport, renderContent()) : renderContent();
};

FloatingBubble.defaultProps = {
  gap: 24,
  axis: 'y',
  offset: {
    x: -1,
    y: -1,
  },
  teleport: () => document.body,
};

export default FloatingBubble;
