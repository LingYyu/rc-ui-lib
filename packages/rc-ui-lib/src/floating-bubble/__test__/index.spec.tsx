import { render, cleanup, fireEvent } from '@testing-library/react';
import React from 'react';
import { sleep } from '../../../tests/utils';
import TestsEvent from '../../../tests/events';
import { mockGetBoundingClientRect } from '../../utils/dom/mock';
import FloatingBubble from '../FloatingBubble';

describe('FloatingBubble', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'inBrowser', { value: true });
  });
  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });
  it('should render correctly when icon set', async () => {
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    const { container } = render(<FloatingBubble icon="chat-o" teleport={root} />);
    await sleep(10);
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    expect(floatingBubbleEl.style.transform).toEqual(
      `translate3d(${window.innerWidth - 48 - 24}px, ${window.innerHeight - 48 - 24}px, 0)`,
    );
    expect(floatingBubbleEl.querySelector('.van-icon-chat-o')).not.toBeNull();
    expect(container).toMatchSnapshot();
    restore();
  });
  it('should render correctly when offset set', async () => {
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    render(<FloatingBubble icon="chat-o" teleport={root} offset={{ x: 100, y: 100 }} />);
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    await sleep(10);

    expect(floatingBubbleEl.style.transform).toEqual(`translate3d(${100}px, ${100}px, 0)`);

    await TestsEvent.triggerDrag(floatingBubbleEl, [24, 1000]);
    expect(floatingBubbleEl.style.transform).toEqual(
      `translate3d(${100}px, ${window.innerHeight - 48 - 24}px, 0)`,
    );
    restore();
  });
  it('should only y axis direction move when axis is default', async () => {
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    render(<FloatingBubble icon="chat-o" teleport={root} />);
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    await TestsEvent.triggerDrag(floatingBubbleEl, [-100, -100]);
    expect(floatingBubbleEl.style.transform).toEqual(
      `translate3d(${window.innerWidth - 48 - 24}px, 24px, 0)`,
    );
    restore();
  });
  it('should only x axis direction move when axis is x', async () => {
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    render(<FloatingBubble icon="chat-o" axis="x" teleport={root} />);
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    await TestsEvent.triggerDrag(floatingBubbleEl, [2000, -100]);
    expect(floatingBubbleEl.style.transform).toEqual(
      `translate3d(${window.innerWidth - 24}px, ${window.innerHeight - 48 - 24}px, 0)`,
    );
    restore();
  });
  it('should free direction move when axis is "xy"', async () => {
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    render(<FloatingBubble icon="chat-o" axis="xy" teleport={root} />);
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    await TestsEvent.triggerDrag(floatingBubbleEl, [-1000, 24]);
    expect(floatingBubbleEl.style.transform).toEqual(
      `translate3d(${window.innerWidth - 48 - 24}px, ${window.innerHeight - 48 - 24}px, 0)`,
    );
    restore();
  });
  it('should magnetic to x axios when magnetic is "x" ', async () => {
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    render(<FloatingBubble icon="chat-o" axis="xy" magnetic="x" teleport={root} />);
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    await TestsEvent.triggerDrag(floatingBubbleEl, [100, 100]);
    await sleep(100);
    expect(floatingBubbleEl.style.transform).toEqual(`translate3d(24px, 100px, 0)`);
    restore();
  });
  it('should magnetic to y axios when magnetic is "y" ', async () => {
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    render(<FloatingBubble icon="chat-o" axis="xy" magnetic="y" teleport={root} />);
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    await TestsEvent.triggerDrag(floatingBubbleEl, [100, 100]);
    await sleep(100);
    expect(floatingBubbleEl.style.transform).toEqual(`translate3d(100px, 24px, 0)`);
    restore();
  });
  it('should emit click when click wrapper', async () => {
    const onClick = jest.fn();
    const root = document.createElement('div');
    render(
      <FloatingBubble icon="chat-o" axis="xy" magnetic="y" teleport={root} onClick={onClick} />,
    );
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    await fireEvent.click(floatingBubbleEl);
    expect(onClick).toHaveBeenCalled();
  });
  it('should offsetchange when touchend', async () => {
    const onClick = jest.fn();
    const onOffsetChange = jest.fn();
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    render(
      <FloatingBubble
        icon="chat-o"
        axis="xy"
        magnetic="y"
        teleport={root}
        onClick={onClick}
        onOffsetChange={onOffsetChange}
      />,
    );
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;

    await TestsEvent.triggerDrag(floatingBubbleEl, [10, 10]);
    expect(onClick).not.toHaveBeenCalled();
    await sleep(10);
    expect(onOffsetChange).toHaveBeenCalled();
    restore();
  });
  it('should lock drag when axis is "lock"', async () => {
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    render(<FloatingBubble icon="chat-o" axis="lock" magnetic="y" teleport={root} />);
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    await TestsEvent.triggerDrag(floatingBubbleEl, [0, 0]);
    expect(floatingBubbleEl.style.transform).toEqual(
      `translate3d(${window.innerWidth - 48 - 24}px, ${window.innerHeight - 48 - 24}px, 0)`,
    );
    restore();
  });
  it('should not move when drag distance belove TAP_OFFSET"', async () => {
    const restore = mockGetBoundingClientRect({ width: 48, height: 48 });
    const root = document.createElement('div');
    render(<FloatingBubble icon="chat-o" axis="x" magnetic="y" teleport={root} />);
    const floatingBubbleEl = root.querySelector<HTMLDivElement>('.rc-floating-bubble')!;
    await TestsEvent.triggerDrag(floatingBubbleEl, [4, 4]);
    expect(floatingBubbleEl.style.transform).toEqual(
      `translate3d(${window.innerWidth - 48 - 24}px, ${window.innerHeight - 48 - 24}px, 0)`,
    );
    restore();
  });

  it('should teleport body when default', async () => {
    const { container } = render(<FloatingBubble icon="chat-o" axis="xy" magnetic="y" />);
    expect(container.parentNode).toEqual(document.body);
    expect(container).toMatchSnapshot();
  });
});
