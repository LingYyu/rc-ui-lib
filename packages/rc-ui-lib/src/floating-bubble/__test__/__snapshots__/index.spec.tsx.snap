// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingBubble should render correctly when icon set 1`] = `<div />`;

exports[`FloatingBubble should teleport body when default 1`] = `<div />`;
